bpo-30520: Implemented pickling for loggers.

vsajip opened the pull request.

vsajip committed:
Implemented pickling for loggers.

vsajip committed:
Implemented pickling for loggers.

pitrou commented on Doc/library/logging.rst:
```diff
@@ -323,6 +323,8 @@ is the module's name in the Python package namespace.
 
       .. versionadded:: 3.2
 
+   .. versionadded:: 3.7
```
That should probably be a `versionchanged`.

vsajip replied:
Done.

vsajip committed:
versionadded -> versionchanged.

vsajip committed:
versionadded -> versionchanged.

serhiy-storchaka commented on Lib/test/test_logging.py:
```diff
@@ -4086,6 +4086,13 @@ def test_invalid_names(self):
+    def test_pickling(self):
+        for name in ('', 'root', 'foo', 'foo.bar', 'baz.bar'):
+            logger = logging.getLogger(name)
+            s = pickle.dumps(logger)
```
Please test it for all protocols:
```
for proto in range(pickle.HIGHEST_PROTOCOL + 1):
    s = pickle.dumps(logger, proto)
```

serhiy-storchaka commented on Lib/logging/__init__.py:
```diff
@@ -1583,6 +1590,9 @@ def __init__(self, level):
         Logger.__init__(self, "root", level)
 
+    def __reduce__(self):
+        return getLogger, ('',)
```
Maybe without arguments at all?
```
return getLogger, ()
```

serhiy-storchaka commented on Lib/logging/__init__.py:
```diff
@@ -1570,6 +1570,13 @@ def __repr__(self):
+    def __reduce__(self):
+        # In general, only the root logger will not be accessible via its name.
+        # However, the root logger's class has its own __reduce__ method.
+        if getLogger(self.name) is not self:
```
This implicitly creates the global logger with name `self.name`. Is it okay?

vsajip replied:
I don't see a problem with it.

serhiy-storchaka commented on Lib/logging/__init__.py:
```diff
@@ -1570,6 +1570,13 @@ def __repr__(self):
+        # In general, only the root logger will not be accessible via its name.
+        # However, the root logger's class has its own __reduce__ method.
+        if getLogger(self.name) is not self:
+            raise pickle.PicklingError('logger cannot be pickled')
```
The `pickle` module is used only here. It isn't used if pickling is not involved. May be make an exception from PEP 8 and import it locally?

serhiy-storchaka commented on Lib/test/test_logging.py:
```diff
@@ -4086,6 +4086,13 @@ def test_invalid_names(self):
         self.assertRaises(TypeError, logging.getLogger, any)
         self.assertRaises(TypeError, logging.getLogger, b'foo')
 
+    def test_pickling(self):
```
`__reduce__` is used also in the `copy` module. Maybe add tests for `copy.copy()` and `copy.deepcopy()`? Even unpickleable loggers can be copied if add special `__copy__` and `__deepcopy__` methods or register the classes with `copyreg.pickle()`.

vsajip replied:
Loggers are singletons, so `copy` operations shouldn't actually make copies, anyway. Seems like a separate issue (or at least use case).

serhiy-storchaka commented on Lib/test/test_logging.py:
```diff
@@ -4086,6 +4086,13 @@ def test_invalid_names(self):
         self.assertRaises(TypeError, logging.getLogger, b'foo')
 
+    def test_pickling(self):
+        for name in ('', 'root', 'foo', 'foo.bar', 'baz.bar'):
```
Is `self.logger` pickleable? If not, it may make sense to add an explicit test for this.

vsajip replied:
No, it's not pickleable, as it's created using `Logger.__init__()`, which users are not supposed to use.

serhiy-storchaka replied:
It is created via `Logger` constructor, not via `getLogger()`. `logging.Logger('blah')` is not `logging.getLogger('blah')`.

vsajip committed:
Addressed Serhiy's review comments.

vsajip committed:
Addressed Serhiy's review comments.

serhiy-storchaka reviewed (approved):
LGTM.

serhiy-storchaka commented:
Ah, just please add a Misc/NEWS entry.

zooba committed:
bpo-30557: faulthandler now correctly filters and displays exception codes on Windows (#1924)

* bpo-30557: faulthandler now correctly filters and displays exception codes on Windows

* Adds test for non-fatal exceptions.

* Adds bpo number to comment.

zware committed:
bpo-30417: Disable 'cpu' and 'tzdata' resources on Travis (GH-1928)

Also weakens the 'should this be run?' regex to allow all builds when .travis.yml changes.

zware committed:
Install dependencies in Travis OSX build (GH-1952)

oz123 committed:
bpo-30095: Make CSS classes used by calendar.HTMLCalendar customizable (GH-1439)

Several class attributes have been added to calendar.HTMLCalendar that allow customization of the CSS classes used in the resulting HTML. This can be done by subclasses HTMLCalendar and overwriting those class attributes (Patch by Oz Tiram).

alex committed:
Simplify X.509 extension handling code (#1855)

* Simplify X.509 extension handling code

The previous implementation had grown organically over time, as OpenSSL's API evolved.

* Delete even more code

vsajip committed:
Merge branch 'pickle-logger' of github.com:vsajip/cpython into pickle-logger

vsajip committed:
Added NEWS entry.
